### PR TITLE
CUDA Loops: move address computation into policy, make policy.load load all arguments

### PR DESCRIPTION
**Pull request by zasdfgbnm**
So that in the future we can make policy accept an offset calculator in its constructor for the support of non-contiguous tensors.

The `elementwise_kernel_helper` is now very general and it can handle any cases:

```C++
template<typename func_t, typename policy_t>
__device__ inline void elementwise_kernel_helper(func_t f, policy_t policy) {
  using traits = function_traits<func_t>;
  using return_t = typename traits::result_type;
  using args_t = typename traits::ArgsTuple;

  int idx = blockIdx.x;

  return_t results[thread_work_size];
  cuda9::workaround::enable_default_constructor<args_t> args_[thread_work_size];
  args_t *args = reinterpret_cast<args_t *>(&args_);

  // load
  policy.load(args, idx);

  // compute
  #pragma unroll
  for (int i = 0; i < thread_work_size; i++) {
    if (policy.check_inbounds(i)) {
      results[i] = c10::guts::apply(f, args[i]);
    }
  }

  // store
  policy.store(results, idx);
}
```

### CODE REVIEW

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 1]
```diff
@@ -145,118 +145,23 @@ invoke(const func_t &f, char *const C10_RESTRICT data[], const index_t strides[]
 // See the note for namespace legacy above.
 namespace modern {
 
-namespace detail {
-
-// The `pointers` converts std::tuple<T1, T2, ....> to std::tuple<T1*, T2*, ....>
-
-template <typename T>
-struct pointers_helper {};
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: `pointers` is no longer needed, therefore deleted

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 2]
```diff
@@ -145,118 +145,23 @@ invoke(const func_t &f, char *const C10_RESTRICT data[], const index_t strides[]
...
-
-template <typename... types>
-struct pointers_helper<std::tuple<types...>> {
-  using type = std::tuple<types *...>;
-};
-
-template <typename T>
-using pointers = typename pointers_helper<T>::type;
-
-// What does the `static_unroll` do?
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: `static_unroll` is moved to `MemoryAccess.cuh`

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 3]
```diff
@@ -145,118 +145,23 @@ invoke(const func_t &f, char *const C10_RESTRICT data[], const index_t strides[]
...
-};
-
-template<template<int i> typename func, int end>
-struct static_unroll<func, end, end> {
-  template<typename... Args>
-  static inline C10_HOST_DEVICE void with_args(Args... args) {}
-};
-
-template<int i>
-struct can_vectorize_up_to_helper {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: Moved to `MemoryAccess.cuh`

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 4]
```diff
@@ -145,118 +145,23 @@ invoke(const func_t &f, char *const C10_RESTRICT data[], const index_t strides[]
...
-  // We need to get the type for each argument of `func_t`, this can only
-  // be done at compile time.
-  static_unroll<can_vectorize_up_to_helper, arity>::with_args(result, pointers, traits());
-  return result;
-}
-
-}  // namespace detail
-
-template<int i>
-struct compute_base_ptrs {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: Absorbed into `load_with_policy` and moved to `MemoryAccess.cuh`

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 5]
```diff
@@ -267,23 +169,28 @@ __device__ inline void elementwise_kernel_helper(func_t f, array_t data, policy_
...
+    elementwise_kernel_helper(f, typename memory::policies::unroll<array_t>(data, remaining));
   } else {  // if this block has a full `block_work_size` data to handle, use vectorized memory access
-    elementwise_kernel_helper(f, data, typename memory::policies::template vectorized<vec_size>());
+    elementwise_kernel_helper(f, typename memory::policies::template vectorized<vec_size, array_t>(data));
   }
 }
 
+template<typename func_t, typename array_t>
+C10_LAUNCH_BOUNDS_1(num_threads)
+__global__ void unrolled_elementwise_kernel(int N, func_t f, array_t data) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh]: This kernel will be used to handle all other cases: non-contiguous, and dynamic casting

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 6]
```diff
@@ -4,12 +4,61 @@
...
+
+template<int i>
+struct load_with_policy {
+  template <typename args_t, typename policy_t>
+  static __device__ void apply(policy_t &self, args_t *args, int idx) {
+    using arg_t = std::tuple_element_t<i, args_t>;
+    // `data` hold the data_ptr for tensors [output, input0, input1, ...], so we
+    // need a +1 offset to get the input
+    auto ptr = reinterpret_cast<arg_t *>(self.data[i + 1]) + block_work_size * idx;
+    auto args_accessor = [&args] __device__ (int index) -> arg_t & { return std::get<i>(args[index]); };
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh]: another nit: too many indices, call it thread_unroll_idx?

**Review by ngimel — APPROVED** [review]
That's great work! Unfortunately, I've let it grow too old, so you need to rebase, sorry about that. 
edit: nm, I see you merged master couple days back, hopefully should be good.

**Comment by zasdfgbnm** [comment]
rebased

**Pull request closed without merging**